lib/Dialect/MSFT/MSFTOps.cpp

```diff
@@ -0,0 +1,60 @@
+//===- MSFTOps.cpp - Implement MSFT dialect operations --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the MSFT dialect operations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/MSFT/MSFTOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace msft;
+
+/// Lookup the module or extmodule for the symbol.  This returns null on
+/// invalid IR.
+Operation *InstanceOp::getReferencedModule() {
+  auto topLevelModuleOp = (*this)->getParentOfType<ModuleOp>();
+  if (!topLevelModuleOp)
```

Comment:
Also curious: why return nullptr, and not assert / return error if the referenced module doesn't exist? It seems unintended to have an instance of a non-existent module.

Reply:
IMO this could be an assertion.

Reply:
This is what `hw.instance` does here. I kept it for consistency.

Reply:
Is there a specific reason? Are instances going to exist without a top-level module in either this dialect or `HW`?

For more context, looking at git blame [here](https://github.com/llvm/circt/commit/3f7ed33578f6068e53ae8a06030af40a1b8136c5#diff-1d225e2b997a8d4507168bd13ef7697369a37fa20aebfb321b8b5ae1e18aa73cR262-R268), it seems this was done initially by @lattner for `hw.instance`.